src/main/ank/1.7-functors-Part2.md

```diff
@@ -0,0 +1,170 @@
+fun <A, B> fmap(f: (A) -> B) =
+    { opt: Option<A> ->
+        opt.fold(
+            ifEmpty = { None },
+            ifSome = { a -> Some(f(a)) }
+        )
+    }
+```
+................
+```cpp
+template<class A, class B>
+optional<B> fmap(std::function<B(A)> f, optional<A> opt) {
+    if (!opt.isValid())
+        return optional<B>{};
+    else
+        return optional<B>{ f(opt.val()) };
+}
+```
+```kotlin:ank:playground
+import arrow.core.Option
+import arrow.core.None
+import arrow.core.Some
+import arrow.core.extensions.option.foldable.fold
+
+fun <A, B> fmap(f: (A) -> B, opt: Option<A>): Option<B> =
+    opt.fold(ifEmpty = { None },
+        ifSome = { a -> Some(f(a)) })
+        
+```
+................
+```Haskell
+auto g = fmap(f);
+```
+```kotlin
+auto(g) == fmap(f)
+```
+................
+```Haskell
+class Eq a where
+    (==) :: a -> a -> Bool
+```
+```kotlin:ank:silent
+interface Eq<F> {
+  fun F.eqv(b: F): Boolean
+
+  fun F.neqv(b: F): Boolean =
+    !eqv(b)
+}
+```
+................
+```Haskell
+data Point = Pt Float Float
+```
+```kotlin:ank:silent
+data class Pt(val x: Float, val y: Float) {
+    companion object
+}
+```
+................
+```Haskell
+instance Eq Point where
+(Pt x y) == (Pt x' y') = x == x' && y == y'
+```
+```kotlin:ank:silent
+import arrow.extension
+
+@extension
+interface PtEq : Eq<Pt> {
+    override fun Pt.eqv(b: Pt): Boolean = x == b.x && y == b.y
+}
+```
+................
+```Haskell
+class Functor f where
+    fmap :: (a -> b) -> f a -> f b
+```
+```kotlin
+interface Functor<F> {
+    //…
+    fun <A, B> Kind<F, A>.map(f: (A) -> B): Kind<F, B>
+}
+```
+................
+```Haskell
+instance Functor Maybe where
+    fmap _ Nothing = Nothing
+    fmap f (Just x) = Just (f x)
+```
+```kotlin
+sealed class Option<out A> {
+    //…
+    inline fun <B> map(f: (A) -> B): Option<B> =
+        flatMap { a -> Some(f(a)) }
+
+    inline fun <B> flatMap(f: (A) -> Option<B>): Option<B> =
+        when (this) {
+              is None -> this
+              is Some -> f(t).fix()
+            }
+}
+```
+................
+```Haskell
+data List a = Nil | Cons a (List a)
+```
+```kotlin:ank:playground
+sealed class List<out A> {
+    object Nil : List<Nothing>()
+    data class Cons<out A>(val head: A, val tail: List<A>) : List<A>()
+}
+```
+................
+```Haskell
+fmap :: (a -> b) -> (List a -> List b)
+```ddd
+```kotlin
+fun <A, B> fmap(): ((A) -> B) -> ((Kind<ForListK, A>) -> Kind<ForListK, B>)
+```
+................
+```Haskell
+fmap f (Cons x t) = Cons (f x) (fmap f t)
+```
+```kotlin:ank:playground
+import arrow.data.ForListK
+import arrow.data.extensions.listk.functor.functor
+import arrow.data.ListK
+import arrow.Kind
+
+fun <A, B> fmap(): ((A) -> B, Kind<ForListK, A>) -> Kind<ForListK, B> =
+    { f, list -> ListK.functor().lift(f)(list) }
+```
+................
+```Haskell
+instance Functor List where
+    fmap _ Nil = Nil
+    fmap f (Cons x t) = Cons (f x) (fmap f t)
+```
+```kotlin:ank:playground
```

Comment:
I'm not sure if this @extension will work on the playground.

Reply:
Yes I forgot the import